rpcs3/Emu/RSX/Overlays/overlay_osk.cpp

```diff
@@ -0,0 +1,563 @@
+﻿#include "stdafx.h"
+#include "overlays.h"
+
+namespace rsx
+{
+	namespace overlays
+	{
+		void osk_dialog::Close(bool ok)
+		{
+			if (on_osk_close)
+			{
+				if (ok)
+				{
+					on_osk_close(CELL_MSGDIALOG_BUTTON_OK);
+				}
+				else
+				{
+					on_osk_close(CELL_MSGDIALOG_BUTTON_ESCAPE);
+				}
+			}
+
+			m_visible = false;
+			close();
+		}
+
+		void osk_dialog::initialize_layout(const std::vector<grid_entry_ctor>& layout, const std::string& title, const std::string& initial_text)
+		{
+			const u32 cell_count = num_rows * num_columns;
+
+			m_grid.resize(cell_count);
+			u32 index = 0;
+
+			for (const auto& props : layout)
+			{
+				for (u32 c = 0; c < props.num_cell_hz; ++c)
+				{
+					const auto row = (index / num_columns);
+					const auto col = (index % num_columns);
+					verify(HERE), row < num_rows && col < num_columns;
+
+					auto &_cell = m_grid[index++];
+					_cell.pos = { col * cell_size_x, row * cell_size_y };
+					_cell.backcolor = props.color;
+					_cell.callback = props.callback;
+					_cell.outputs = props.outputs;
+					_cell.selected = 0;
+
+					num_layers = std::max<u32>(num_layers, _cell.outputs.size());
+
+					if (LIKELY(props.num_cell_hz == 1))
+					{
+						_cell.flags = border_flags::default_cell;
+					}
+					else if (c == 0)
+					{
+						// Leading cell
+						_cell.flags = border_flags::start_cell;
+					}
+					else if (c == (props.num_cell_hz - 1))
+					{
+						// Last cell
+						_cell.flags = border_flags::end_cell;
+					}
+					else
+					{
+						// Middle cell
+						_cell.flags = border_flags::middle_cell;
+					}
+				}
+			}
+
+			verify(HERE), num_layers;
+
+			selected_x = selected_y = selected_z = 0;
+			m_grid[0].selected = true;
+			m_update = true;
+
+			m_background.set_size(1280, 720);
+			m_background.back_color.a = 0.8f;
+
+			// Place elements with absolute positioning
+			u16 frame_w = u16(num_columns * cell_size_x);
+			u16 frame_h = u16(num_rows * cell_size_y) + 80;
+			u16 frame_x = (1280 - frame_w) / 2;
+			u16 frame_y = (720 - frame_h) / 2;
+
+			m_frame.set_pos(frame_x, frame_y);
+			m_frame.set_size(frame_w, frame_h);
+			m_frame.back_color = { 0.2f, 0.2f, 0.2f, 1.f };
+
+			m_title.set_pos(frame_x, frame_y);
+			m_title.set_size(frame_w, 30);
+			m_title.set_text(title);
+			m_title.set_padding(15, 0, 5, 0);
+			m_title.back_color.a = 0.f;
+
+			m_preview.set_pos(frame_x, frame_y + 30);
+			m_preview.set_size(frame_w, 40);
+			m_preview.set_text(initial_text.empty()? "[Enter Text]" : initial_text);
```

Comment:
Maybe also set the default preview when the Text field is empty (up for debate)

Reply:
Hint text color changed. If needed everytime the field is empty, its easy to add again.